<a href="https://colab.research.google.com/github/Shruttts/Movie-Recommendation-Model/blob/main/MovieRecommenderSystem.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [3]:
import numpy as np
import pandas as pd

In [4]:
credit=pd.read_csv('/content/tmdb_5000_credits.csv')
movie=pd.read_csv('/content/tmdb_5000_movies.csv')

In [5]:
movie.head(1)

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                      homepage     id  \
0  http://www.avatarmovie.com/  19995   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   

  original_title                                           overview  \
0         Avatar  In the 22nd century, a paraplegic Marine is di...   

   popularity                               production_companies  \
0  150.437577  [{"name": "Ingenious Film Partners", "id": 289...   

                                production_countries release_date     revenue  \
0  [{"iso_3166_1": "US", "name": "United States o...   2009-12-10  2787965087   

   runtime                                   spoken_languages    status  \
0    162.0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   

                       tagline   title  vote_average  vote_count  
0  Enter the World of Pandora.  Avatar           7.2       11800

In [6]:
credit.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

In [7]:
credit.shape

(4803, 4)

In [8]:
movie.shape

(4803, 20)

In [9]:
movie=movie.merge(credit, on ='title')

In [10]:
movie.shape

(4809, 23)

In [11]:
#fro recommendation we need to create a tag for every movie so will check where it is importatnt so that we can create a tag

1. genres
2. id
3. keyword (this are like a tags)
4. title
5. overview
6. cast
7. crew

In [12]:
movie['original_language'].value_counts()

original_language
en    4510
fr      70
es      32
zh      27
de      27
hi      19
ja      16
it      14
ko      12
cn      12
ru      11
pt       9
da       7
sv       5
nl       4
fa       4
th       3
he       3
ta       2
cs       2
ro       2
id       2
ar       2
vi       1
sl       1
ps       1
no       1
ky       1
hu       1
pl       1
af       1
nb       1
tr       1
is       1
xx       1
te       1
el       1
Name: count, dtype: int64

In [13]:
movie.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4809 entries, 0 to 4808
Data columns (total 23 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   budget                4809 non-null   int64  
 1   genres                4809 non-null   object 
 2   homepage              1713 non-null   object 
 3   id                    4809 non-null   int64  
 4   keywords              4809 non-null   object 
 5   original_language     4809 non-null   object 
 6   original_title        4809 non-null   object 
 7   overview              4806 non-null   object 
 8   popularity            4809 non-null   float64
 9   production_companies  4809 non-null   object 
 10  production_countries  4809 non-null   object 
 11  release_date          4808 non-null   object 
 12  revenue               4809 non-null   int64  
 13  runtime               4807 non-null   float64
 14  spoken_languages      4809 non-null   object 
 15  status               

In [14]:
movie=movie[['movie_id','title','overview','genres','keywords','cast','crew']]

In [15]:
movie.head(1)

movie_id   title                                           overview  \
0     19995  Avatar  In the 22nd century, a paraplegic Marine is di...   

                                              genres  \
0  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                            keywords  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...

now we only want 3 coloumns
1. movie_id
2. title
3. tag

Now we will do data preprocessing like we going to make it in a sameformat



To check missing data

In [16]:
movie.isnull().sum()

movie_id    0
title       0
overview    3
genres      0
keywords    0
cast        0
crew        0
dtype: int64

In [17]:
movie.dropna(inplace=True)

In [18]:
movie.shape

(4806, 7)

To check duplicate data

In [19]:
movie.duplicated().sum()

0

Now its time to make everything in the same format

In [20]:
movie.iloc[0].genres
#this is a list of dictionary

'[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]'

creating helper function to make it in a format oflist of names only

In [21]:
import ast


ast.literal_eval() this function from ast module in python which stands for Abstract Syntax Trees will convert the list of string to the list

In [22]:
def convert(obj):
  l=[]
  for i in ast.literal_eval(obj):
   l.append(i['name'])

  return l

In [23]:
movie['genres']=movie['genres'].apply(convert)

In [24]:
movie.iloc[0].keywords

'[{"id": 1463, "name": "culture clash"}, {"id": 2964, "name": "future"}, {"id": 3386, "name": "space war"}, {"id": 3388, "name": "space colony"}, {"id": 3679, "name": "society"}, {"id": 3801, "name": "space travel"}, {"id": 9685, "name": "futuristic"}, {"id": 9840, "name": "romance"}, {"id": 9882, "name": "space"}, {"id": 9951, "name": "alien"}, {"id": 10148, "name": "tribe"}, {"id": 10158, "name": "alien planet"}, {"id": 10987, "name": "cgi"}, {"id": 11399, "name": "marine"}, {"id": 13065, "name": "soldier"}, {"id": 14643, "name": "battle"}, {"id": 14720, "name": "love affair"}, {"id": 165431, "name": "anti war"}, {"id": 193554, "name": "power relations"}, {"id": 206690, "name": "mind and soul"}, {"id": 209714, "name": "3d"}]'

In [25]:
def convert5(obj):
  l=[]
  counter=0
  for i in ast.literal_eval(obj):
   l.append(i['name'])
   counter+=1
   if counter==5:
    break;
  return l

In [26]:
movie['keywords']

0       [{"id": 1463, "name": "culture clash"}, {"id":...
1       [{"id": 270, "name": "ocean"}, {"id": 726, "na...
2       [{"id": 470, "name": "spy"}, {"id": 818, "name...
3       [{"id": 849, "name": "dc comics"}, {"id": 853,...
4       [{"id": 818, "name": "based on novel"}, {"id":...
                              ...                        
4804    [{"id": 5616, "name": "united states\u2013mexi...
4805                                                   []
4806    [{"id": 248, "name": "date"}, {"id": 699, "nam...
4807                                                   []
4808    [{"id": 1523, "name": "obsession"}, {"id": 224...
Name: keywords, Length: 4806, dtype: object

In [27]:
movie['keywords']=movie['keywords'].apply(convert)

In [28]:
movie.iloc[0].cast

'[{"cast_id": 242, "character": "Jake Sully", "credit_id": "5602a8a7c3a3685532001c9a", "gender": 2, "id": 65731, "name": "Sam Worthington", "order": 0}, {"cast_id": 3, "character": "Neytiri", "credit_id": "52fe48009251416c750ac9cb", "gender": 1, "id": 8691, "name": "Zoe Saldana", "order": 1}, {"cast_id": 25, "character": "Dr. Grace Augustine", "credit_id": "52fe48009251416c750aca39", "gender": 1, "id": 10205, "name": "Sigourney Weaver", "order": 2}, {"cast_id": 4, "character": "Col. Quaritch", "credit_id": "52fe48009251416c750ac9cf", "gender": 2, "id": 32747, "name": "Stephen Lang", "order": 3}, {"cast_id": 5, "character": "Trudy Chacon", "credit_id": "52fe48009251416c750ac9d3", "gender": 1, "id": 17647, "name": "Michelle Rodriguez", "order": 4}, {"cast_id": 8, "character": "Selfridge", "credit_id": "52fe48009251416c750ac9e1", "gender": 2, "id": 1771, "name": "Giovanni Ribisi", "order": 5}, {"cast_id": 7, "character": "Norm Spellman", "credit_id": "52fe48009251416c750ac9dd", "gender": 

In [29]:
def convert3(obj):
  l=[]
  counter=0
  for i in ast.literal_eval(obj):
   l.append(i['name'])
   counter+=1
   if counter==3:
    break;

  return l

In [30]:
movie['cast']=movie['cast'].apply(convert3)

In [31]:
movie['cast']

0        [Sam Worthington, Zoe Saldana, Sigourney Weaver]
1           [Johnny Depp, Orlando Bloom, Keira Knightley]
2            [Daniel Craig, Christoph Waltz, Léa Seydoux]
3            [Christian Bale, Michael Caine, Gary Oldman]
4          [Taylor Kitsch, Lynn Collins, Samantha Morton]
                              ...                        
4804    [Carlos Gallardo, Jaime de Hoyos, Peter Marqua...
4805         [Edward Burns, Kerry Bishé, Marsha Dietlein]
4806           [Eric Mabius, Kristin Booth, Crystal Lowe]
4807            [Daniel Henney, Eliza Coupe, Bill Paxton]
4808    [Drew Barrymore, Brian Herzlinger, Corey Feldman]
Name: cast, Length: 4806, dtype: object

In [32]:
movie.iloc[0].crew

'[{"credit_id": "52fe48009251416c750aca23", "department": "Editing", "gender": 0, "id": 1721, "job": "Editor", "name": "Stephen E. Rivkin"}, {"credit_id": "539c47ecc3a36810e3001f87", "department": "Art", "gender": 2, "id": 496, "job": "Production Design", "name": "Rick Carter"}, {"credit_id": "54491c89c3a3680fb4001cf7", "department": "Sound", "gender": 0, "id": 900, "job": "Sound Designer", "name": "Christopher Boyes"}, {"credit_id": "54491cb70e0a267480001bd0", "department": "Sound", "gender": 0, "id": 900, "job": "Supervising Sound Editor", "name": "Christopher Boyes"}, {"credit_id": "539c4a4cc3a36810c9002101", "department": "Production", "gender": 1, "id": 1262, "job": "Casting", "name": "Mali Finn"}, {"credit_id": "5544ee3b925141499f0008fc", "department": "Sound", "gender": 2, "id": 1729, "job": "Original Music Composer", "name": "James Horner"}, {"credit_id": "52fe48009251416c750ac9c3", "department": "Directing", "gender": 2, "id": 2710, "job": "Director", "name": "James Cameron"},

In [33]:
def convert(obj):
  l=[]

  for i in ast.literal_eval(obj):
    if i['job']=='Director':
       l.append(i['name'])
       break
  return l

In [34]:
movie['crew']=movie['crew'].apply(convert)

In [35]:
movie.head(1)

movie_id   title                                           overview  \
0     19995  Avatar  In the 22nd century, a paraplegic Marine is di...   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   

                                            keywords  \
0  [culture clash, future, space war, space colon...   

                                               cast             crew  
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]  [James Cameron]

In [36]:
movie['overview'][0]

'In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization.'

.apply(lambda x: x.split()) with the help of this function string is converted into list of words

In [37]:
movie['overview'].apply(lambda x: x.split())

0       [In, the, 22nd, century,, a, paraplegic, Marin...
1       [Captain, Barbossa,, long, believed, to, be, d...
2       [A, cryptic, message, from, Bond’s, past, send...
3       [Following, the, death, of, District, Attorney...
4       [John, Carter, is, a, war-weary,, former, mili...
                              ...                        
4804    [El, Mariachi, just, wants, to, play, his, gui...
4805    [A, newlywed, couple's, honeymoon, is, upended...
4806    ["Signed,, Sealed,, Delivered", introduces, a,...
4807    [When, ambitious, New, York, attorney, Sam, is...
4808    [Ever, since, the, second, grade, when, he, fi...
Name: overview, Length: 4806, dtype: object

In [38]:
movie['cast'].apply(lambda x:[i.replace(" ","") for i in x])
movie['keywords'].apply(lambda x:[i.replace(" ","") for i in x])
movie['crew'].apply(lambda x:[i.replace(" ","") for i in x])
movie['genres'].apply(lambda x:[i.replace(" ","") for i in x])

0       [Action, Adventure, Fantasy, ScienceFiction]
1                       [Adventure, Fantasy, Action]
2                         [Action, Adventure, Crime]
3                   [Action, Crime, Drama, Thriller]
4                [Action, Adventure, ScienceFiction]
                            ...                     
4804                       [Action, Crime, Thriller]
4805                               [Comedy, Romance]
4806               [Comedy, Drama, Romance, TVMovie]
4807                                              []
4808                                   [Documentary]
Name: genres, Length: 4806, dtype: object

In [39]:
movie.head(1)

movie_id   title                                           overview  \
0     19995  Avatar  In the 22nd century, a paraplegic Marine is di...   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   

                                            keywords  \
0  [culture clash, future, space war, space colon...   

                                               cast             crew  
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]  [James Cameron]

In [40]:
movie['tags'] = (
    movie['overview'].apply(str) + " " +
    movie['genres'].apply(str) + " " +
    movie['cast'].apply(str) + " " +
    movie['crew'].apply(str) + " " +
    movie['keywords'].apply(str)
)

In [41]:
movie

movie_id                                     title  \
0        19995                                    Avatar   
1          285  Pirates of the Caribbean: At World's End   
2       206647                                   Spectre   
3        49026                     The Dark Knight Rises   
4        49529                               John Carter   
...        ...                                       ...   
4804      9367                               El Mariachi   
4805     72766                                 Newlyweds   
4806    231617                 Signed, Sealed, Delivered   
4807    126186                          Shanghai Calling   
4808     25975                         My Date with Drew   

                                               overview  \
0     In the 22nd century, a paraplegic Marine is di...   
1     Captain Barbossa, long believed to be dead, ha...   
2     A cryptic message from Bond’s past sends him o...   
3     Following the death of District Attorney Harve...   
4     John Carter is a war-weary, former military ca...   
...                                                 ...   
4804  El Mariachi just wants to play his guitar and ...   
4805  A newlywed couple's honeymoon is upended by th...   
4806  "Signed, Sealed, Delivered" introduces a dedic...   
4807  When ambitious New York attorney Sam is sent t...   
4808  Ever since the second grade when he first saw ...   

                                             genres  \
0     [Action, Adventure, Fantasy, Science Fiction]   
1                      [Adventure, Fantasy, Action]   
2                        [Action, Adventure, Crime]   
3                  [Action, Crime, Drama, Thriller]   
4              [Action, Adventure, Science Fiction]   
...                                             ...   
4804                      [Action, Crime, Thriller]   
4805                              [Comedy, Romance]   
4806             [Comedy, Drama, Romance, TV Movie]   
4807                                             []   
4808                                  [Documentary]   

                                               keywords  \
0     [culture clash, future, space war, space colon...   
1     [ocean, drug abuse, exotic island, east india ...   
2     [spy, based on novel, secret agent, sequel, mi...   
3     [dc comics, crime fighter, terrorist, secret i...   
4     [based on novel, mars, medallion, space travel...   
...                                                 ...   
4804  [united states–mexico barrier, legs, arms, pap...   
4805                                                 []   
4806  [date, love at first sight, narration, investi...   
4807                                                 []   
4808          [obsession, camcorder, crush, dream girl]   

                                                   cast                 crew  \
0      [Sam Worthington, Zoe Saldana, Sigourney Weaver]      [James Cameron]   
1         [Johnny Depp, Orlando Bloom, Keira Knightley]     [Gore Verbinski]   
2          [Daniel Craig, Christoph Waltz, Léa Seydoux]         [Sam Mendes]   
3          [Christian Bale, Michael Caine, Gary Oldman]  [Christopher Nolan]   
4        [Taylor Kitsch, Lynn Collins, Samantha Morton]     [Andrew Stanton]   
...                                                 ...                  ...   
4804  [Carlos Gallardo, Jaime de Hoyos, Peter Marqua...   [Robert Rodriguez]   
4805       [Edward Burns, Kerry Bishé, Marsha Dietlein]       [Edward Burns]   
4806         [Eric Mabius, Kristin Booth, Crystal Lowe]        [Scott Smith]   
4807          [Daniel Henney, Eliza Coupe, Bill Paxton]        [Daniel Hsia]   
4808  [Drew Barrymore, Brian Herzlinger, Corey Feldman]   [Brian Herzlinger]   

                                                   tags  
0     In the 22nd century, a paraplegic Marine is di...  
1     Captain Barbossa, long believed to be dead, ha...  
2     A cryptic message from Bond’s past sends him o...  
3     Followi

In [42]:
movie['tags']=movie['tags'].apply(lambda x:x.lower())

In [43]:
new_df=movie[['movie_id','title','tags']]

In [44]:
new_df

movie_id                                     title  \
0        19995                                    Avatar   
1          285  Pirates of the Caribbean: At World's End   
2       206647                                   Spectre   
3        49026                     The Dark Knight Rises   
4        49529                               John Carter   
...        ...                                       ...   
4804      9367                               El Mariachi   
4805     72766                                 Newlyweds   
4806    231617                 Signed, Sealed, Delivered   
4807    126186                          Shanghai Calling   
4808     25975                         My Date with Drew   

                                                   tags  
0     in the 22nd century, a paraplegic marine is di...  
1     captain barbossa, long believed to be dead, ha...  
2     a cryptic message from bond’s past sends him o...  
3     following the death of district attorney harve...  
4     john carter is a war-weary, former military ca...  
...                                                 ...  
4804  el mariachi just wants to play his guitar and ...  
4805  a newlywed couple's honeymoon is upended by th...  
4806  "signed, sealed, delivered" introduces a dedic...  
4807  when ambitious new york attorney sam is sent t...  
4808  ever since the second grade when he first saw ...  

[4806 rows x 3 columns]

#vectorization

now i want to find the similarization between 2 text so that it will recommend me other movies too .

[will convert the text into vector]

from sklearn.feature_extraction.text import CountVectorizer:
This imports the CountVectorizer class from Scikit-learn’s feature_extraction.text module. The CountVectorizer is used to convert a collection of text documents (in your case, from new_df['tags']) into a matrix of token counts (word frequencies).

In [45]:
pip install scikit-learn nltk matplotlib

In [46]:
import nltk
from sklearn.feature_extraction.text import CountVectorizer
import matplotlib.pyplot as plt
from nltk.corpus import stopwords

In [47]:
from sklearn.feature_extraction.text import CountVectorizer

cv = CountVectorizer(...):
You are initializing a CountVectorizer object and assigning it to the variable cv.
max_features=5000: This parameter limits the number of features (i.e., the number of unique words) to the top 5000 most frequent words across all the documents in your dataset. This is a way to reduce the dimensionality of the data and focus on the most important words.
stop_words='English': This argument removes common English stop words (like "the", "is", "in", etc.) from the text during the vectorization process. It uses a predefined list of stop words available in Scikit-learn. These words are ignored when the text is transformed into vectors, as they don't contribute much to text analysis.

In [48]:
cv = CountVectorizer(max_features=5000, stop_words='english')

cv.fit_transform(new_df['tags']):

fit_transform() is a method in CountVectorizer that does two things:
fit(): It analyzes the text data in new_df['tags'] to build a vocabulary (a list of unique words) based on the text. The vocabulary will contain only the words that appear in the text, excluding stop words.
transform(): After fitting, it then transforms the text documents into a matrix of word counts. Each document is represented as a vector where each element corresponds to the frequency of a word from the vocabulary.
new_df['tags'] is assumed to be a column in your new_df dataframe containing text data, probably a list of tags or keywords for each document.
.toarray():

This converts the sparse matrix (default output of fit_transform()) into a dense NumPy array. A sparse matrix is memory-efficient and stores only the non-zero elements, but for easier manipulation or inspection, you may want to convert it into a dense array.

In [49]:
cv.fit_transform(new_df['tags']).toarray()

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]])

In [50]:
vectors=cv.fit_transform(new_df['tags']).toarray()

In [51]:
vectors[2]

array([0, 0, 0, ..., 0, 0, 0])

CountVectorizer(max_features=5000, stop_words='English'): Initializes the vectorizer to transform text into word count vectors, limiting to the top 5000 words and removing stop words.
cv.fit_transform(new_df['tags']): This generates a matrix of word counts from the tags column of new_df.
.toarray(): Converts the sparse matrix into a dense NumPy array for easier handling

In [52]:
cv.get_feature_names_out()

array(['000', '007', '10', ..., 'zooey', 'zucker', 'zwick'], dtype=object)

In [53]:
len(cv.get_feature_names_out())

5000

Will do stemming here
Stemming :
['loved', 'loving' , 'isncjcd']
it will give the main word love that is stemming

In [54]:
import nltk
from nltk.stem import PorterStemmer
ps=PorterStemmer()

In [55]:
def stem(text):
         y=[]
         for i in text.split():
           y.append(ps.stem(i))
         return " ".join(y)

In [56]:
ps.stem('loved')

'love'

In [57]:
new_df['tags']=new_df['tags'].apply(stem)

<ipython-input-57-402dd7cb6b10>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df['tags']=new_df['tags'].apply(stem)


here we will not calculate euclidine distance we will calculate cosine distance (distance kam similarity jyada)

In [58]:
from sklearn.metrics.pairwise import cosine_similarity

pahilya movie cha tichya pasunch distance as 4806 movies cha 4806 movie barobar ch distance

In [59]:
similarity=cosine_similarity(vectors)

In [60]:
cosine_similarity(vectors).shape

(4806, 4806)

diagonal element will return the value as 1 (distance with itself)

In [61]:
similarity

array([[1.        , 0.06839856, 0.04916051, ..., 0.03121953, 0.05374308,
        0.        ],
       [0.06839856, 1.        , 0.04259177, ..., 0.04057204, 0.        ,
        0.        ],
       [0.04916051, 0.04259177, 1.        , ..., 0.01944039, 0.08924215,
        0.        ],
       ...,
       [0.03121953, 0.04057204, 0.01944039, ..., 1.        , 0.06375767,
        0.03276488],
       [0.05374308, 0.        , 0.08924215, ..., 0.06375767, 1.        ,
        0.03760222],
       [0.        , 0.        , 0.        , ..., 0.03276488, 0.03760222,
        1.        ]])

In [62]:
similarity[0]

array([1.        , 0.06839856, 0.04916051, ..., 0.03121953, 0.05374308,
       0.        ])

In [63]:
list(enumerate(similarity[0]))

[(0, 0.9999999999999997),
 (1, 0.06839855680567694),
 (2, 0.04916051440083466),
 (3, 0.022496063533292376),
 (4, 0.3083035200691657),
 (5, 0.14933692277237473),
 (6, 0.023537557657892522),
 (7, 0.13111122301831463),
 (8, 0.03934447376823168),
 (9, 0.07466846138618737),
 (10, 0.10375716957991118),
 (11, 0.06765100914917382),
 (12, 0.05510387687779837),
 (13, 0.026669037353133248),
 (14, 0.12605069833265087),
 (15, 0.04335549847620599),
 (16, 0.10366105360784114),
 (17, 0.11047363533322149),
 (18, 0.12617865362880898),
 (19, 0.06806659626466625),
 (20, 0.03342423883661898),
 (21, 0.05019558610576909),
 (22, 0.042601432284230495),
 (23, 0.05510387687779837),
 (24, 0.04518564248465834),
 (25, 0.06653536832815835),
 (26, 0.17013926184468012),
 (27, 0.19597660033920397),
 (28, 0.07113879915305903),
 (29, 0.056055185467121044),
 (30, 0.08812102431690699),
 (31, 0.14933692277237473),
 (32, 0.06622661785325218),
 (33, 0.07755667343294811),
 (34, 0.0),
 (35, 0.179968508266339),
 (36, 0.162221421

In [64]:
sorted(list(enumerate(similarity[0])) , reverse=True,key=lambda x:x[1])[1:6]

[(2405, 0.48074115106715354),
 (1533, 0.4333408306443943),
 (3160, 0.42501658525798836),
 (838, 0.39558756904487125),
 (4336, 0.38949041885226005)]

In [65]:
def recommend(movie):
  movie_index=new_df[new_df['title']==movie].index[0]
  distances=similarity[movie_index]
  movies_list=sorted(list(enumerate(distances)) , reverse=True,key=lambda x:x[1])[1:6]

  for i in movies_list:
    print(new_df.iloc[i[0]].title)

In [68]:
import pickle

In [69]:
pickle.dump(new_df,open('movies.pkl','wb'))